rclcpp/include/rclcpp/experimental/buffers/ring_buffer_implementation.hpp

```diff
@@ -112,7 +115,7 @@ class RingBufferImplementation : public BufferImplementationBase<BufferT>
   size_t read_index_;
   size_t size_;
 
-  std::mutex mutex_;
+  mutable std::recursive_mutex recursive_mutex_;
```

Comment:
this has to be std::recursive_mutex instead mutex? as far as i checked, recursive_mutex takes time more than mutex. And i do not see the necessity so far, could you elaborate the case?

Reply:
I use the recursive mutex because `enqueue()` and `dequeue()` call `next()`, `has_data()`, and `is_full()`, which have locks in them.